source-firestore/discovery.go

```diff
@@ -4,48 +4,92 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"math/rand"
+	"reflect"
 	"sort"
 	"strings"
 	"sync"
+	"time"
 
 	firestore "cloud.google.com/go/firestore"
 	firebase "firebase.google.com/go"
 	"github.com/estuary/connectors/schema_inference"
 	pc "github.com/estuary/flow/go/protocols/capture"
 	pf "github.com/estuary/flow/go/protocols/flow"
+	"github.com/invopop/jsonschema"
 	log "github.com/sirupsen/logrus"
 	"golang.org/x/sync/errgroup"
 	"google.golang.org/api/iterator"
 	"google.golang.org/api/option"
 )
 
 const (
-	// discoverMaxCollectionsPerResource bounds the number of distinct collections
-	// that an invocation of discoverResource will process. If more collections are
-	// discovered which map to a particular resource, the excess will first pile up
-	// in the buffered channel and then after that they will be silently discarded.
-	discoverMaxCollectionsPerResource = 50
-
-	// discoverMaxDocumentsPerCollection bounds the number of documents which will
-	// be fetched by discoverResource from any one collection.
-	discoverMaxDocumentsPerCollection = 100
-
-	// discoverMaxConcurrentWorkers bounds the number of concurrent `discoverResource()`
-	// workers which can proceed at any given time.
-	discoverMaxConcurrentWorkers = 32
+	// discoverSubresourceMinimum and discoverSubresourceProbability work together
+	// to limit the number of documents which we query for subcollections. The logic
+	// here is that if some collection (let's call it "users/*/messages" for example)
+	// has a subcollection ("users/*/messages/*/attachments") we don't need to check
+	// every single message for every user to learn this fact. Furthermore since every
+	// document we check requires a distinct RPC, this could take quite a while to do.
+	//
+	// So instead we can apply some common-sense policy which mimics what a user might
+	// do: check the first N exhaustively and then start skimming through and randomly
```

Comment:
The "first N" part of the heuristic is less about sequential ordering and more about making sure that when a collection only has a few documents every one of them will be checked for possible nested collections.

Implementing "last N" is trickier because we don't know upfront how many documents are in a collection, so we'd have to do some extra work to maintain a list of the last N documents observed from each collection and then launch some workers to check them for sub-collections only after reaching the end of the collection group. Which would definitely work, but in practice it seems like the IDs in most collections are random (and thus non-sequential) unique IDs anyway, so I'm not sure it's worth the extra effort compared to the "randomly sample P% of documents" approach.

(To be clear just in case you missed it, note that we're still feeding every document we receive into schema inference, the activity that this heuristic governs is solely the "does this document have any nested collections" check since doing that for every single document turned out to be the slow part of discovery.)